File: Patterns/Template Pattern/Cpp/Starbuzz/CaffeineBeverageWithHook.cpp
#pragma once

#include <iostream>

#define interface class

using namespace std;

interface CaffeineBeverageWithHook {
public:
    void prepareRecipe() {
        boilWater();
        brew();
        pourInCup();
        if (customerWantsCondiments()) addCondiments();
    }

    virtual void brew() = 0;

    virtual void addCondiments() = 0;

    virtual bool customerWantsCondiments() {
        return true;
    }

    void boilWater() {
        cout << "물 끓이는 중\n";
    }

    void pourInCup() {
        cout << "컵에 따르는 중\n";
    }
};

File: Patterns/Template Pattern/CSharp/Starbuzz/CaffeineBeverageWithHook.cs
using System;

namespace Starbuzz 
{
    public abstract class CaffeineBeverageWithHook
    {
        public void PrepareRecipe() 
        {
            BoilWater();
            Brew();
            PourInCup();
            if (CustomerWantsCondiments()) AddCondiments();
        }

        public abstract void Brew();

        public abstract void AddCondiments();

        public virtual bool CustomerWantsCondiments() => true;

        public void BoilWater() => Console.WriteLine("물 끓이는 중");

        public void PourInCup() => Console.WriteLine("컵에 따르는 중");
    } 
}